### ISSUE TITLE
Admin API?

### ISSUE THREAD

**Issue by brylie** [mail-in-a-box/mailinabox#542]
We are considering how to write software that would provide integration with a mail-in-a-box instance. What are our options for integration? Does mail-in-a-box have an administrative API, such as via REST?

**Comment by JoshData** [comment]
All of the control panel functionality is through a RESTy API:

https://github.com/mail-in-a-box/mailinabox/blob/master/management/daemon.py#L134

Only the Custom DNS part has any documentation, however. (The documentation is within the control panel.)

**Comment by brylie** [comment]
OK, thanks for the quick response @JoshData. I updated the issue to indicate that it is a possible improvement.

**Comment by nikolay** [comment]
BTW, there's [RAML](http://raml.org/) as well...

**Comment by brylie** [comment]
Also, the Linux Foundation is working on an Open API  Initiative, which is [extending the  Swagger specification](https://openapis.org/news/announcement/2015/11/new-collaborative-project-extend-swagger-specification-building-connected).

**Comment by badsyntax** [comment]
hi!  i have made good progress documenting the API within this repo: https://github.com/badsyntax/mailinabox-api (see [mailinabox.yml](https://github.com/badsyntax/mailinabox-api/blob/master/schema/mailinabox.yml)) and i will likely finish documenting all endpoints in the coming days. 

is this something that belongs in the [mailinabox](https://github.com/mail-in-a-box/mailinabox) repo? i am happy to contribute the spec, let me know where it belongs and i'll send a PR when i'm finished.

**Comment by JoshData** [comment]
That's amazing! Yes we should add the mailinabox.yml file here with instructions for what to do with it, and if it auto-generates HTML documentation we can host that on mailinabox.email.

**Comment by badsyntax** [comment]
Cool. I'll send a PR soonish. 

We can indeed generate HTML API docs, is there a particular API docs framework/format you'd like to use? I don't mind doing the work to generate the docs, but i don't have a strong opinion on how the docs should look/behave.

## Option 1 - [swagger-ui](https://github.com/swagger-api/swagger-ui)

This is a dynamic doc generator. It loads the schema and generates the UI with JavaScript.

Example: https://petstore.swagger.io/?url=https://raw.githubusercontent.com/badsyntax/mailinabox-api/master/schema/mailinabox.yml

If we want to customise swagger-ui (like remove the swagger branding, add a mailinabox logo, or adjust the styles to fit in with the website) we can do it via a static html page (or build it as a React app):

<details><summary>Example</summary>

```html
<!DOCTYPE html>
<html>
  <head>
    <title>Swagger</title>
    <meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/3.31.1/swagger-ui.min.css" />
    <style type="text/css">
    /* override styles here */
    </style>
  </head>
  <body>
    <div id="swagger-ui"></div>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/3.31.1/swagger-ui-bundle.min.js"></script>
    <script>
    const ui = SwaggerUIBundle({
        url: "https://raw.githubusercontent.com/badsyntax/mailinabox-api/master/schema/mailinabox.yml",
        dom_id: '#swagger-ui',
        presets: [
          SwaggerUIBundle.presets.apis,
        ],
        layout: "BaseLayout",
        defaultModelsExpandDepth: 0,
        deepLinking: true
      })
    </script>
  </body>
</html>
```
</details>

## Option 2 -  [slate](https://github.com/slatedocs/slate)

Static HTML docs. Seems like a solid/mature tool and is customisable.

Example: https://slatedocs.github.io/slate/ 
Customised example: http://v2apidoc.octobat.com/

## Other options?

There's many tools to convert the spec to various documentation formats, i'm just not that familiar with good API doc frameworks/formats. Can you suggest anything?

**Comment by JoshData** [comment]
My preference would be whatever is easiest to maintain: Having a single-file output (just one HTML page, no static assets) would be convenient. Slate seems to do that?

**Comment by badsyntax** [comment]
Yes it looks like it. Once the slate template is setup correctly, the maintenance burden will be on re-generating the API docs whenever the API spec changes. We don't maintain the generated HTML file, we maintain the workflow used to generated the HTML file. The workflow involves running a script. The spec should be treated as the source of truth, and the API docs should be an artifact of the spec.

Ideally the docs should be generated via CI/CD on changes to master branch but I fully understand the need to keep things simple, in which case, we'll need to remember to manually re-generate the api docs whenever the spec changes. 
How is the website currently deployed? Is it along the line of "git pull"?

**Comment by JoshData** [comment]
That all sounds fantastic. Yes, I git pull the website repo into my own Mail-in-a-Box, where it's served from.

**Comment by badsyntax** [comment]
https://github.com/mail-in-a-box/mailinabox/pull/1804 is ready for review.
